### PR TITLE
#6386 modified create purchase orders from procurement window process.

### PR DESCRIPTION
**Pull request by dragospodariu96**
https://github.com/metasfresh/metasfresh/issues/6386

### CODE REVIEW

File: de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java [hunk 1]
```diff
@@ -66,7 +64,20 @@ public Result processWorkPackage(final I_C_Queue_WorkPackage workPackage, final
...
-		return Services.get(IQueueDAO.class).retrieveItemsSkipMissing(workpackage, I_PMM_PurchaseCandidate.class, ITrx.TRXNAME_ThreadInherited);
+		return Services.get(IQueueDAO.class).retrieveAllItems(workpackage, I_PMM_PurchaseCandidate.class);
+	}
+
+	private List<I_PMM_PurchaseCandidate> getItemsWithQty(List<I_PMM_PurchaseCandidate> candidates)
+	{
+		List<I_PMM_PurchaseCandidate> filteredList = new ArrayList<I_PMM_PurchaseCandidate>();
+		for (I_PMM_PurchaseCandidate candidate : candidates)
+		{
+			if (candidate.getQtyOrdered().compareTo(BigDecimal.ZERO) > 0)
```
Comment (metas-ts) [inline de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java]: * use `BigDecimal.signum()` instead

File: de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java [hunk 2]
```diff
@@ -66,7 +64,20 @@ public Result processWorkPackage(final I_C_Queue_WorkPackage workPackage, final
...
+			{
+				filteredList.add(candidate);
+			}
```
Comment (metas-ts) [inline de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java]: * explicitly log the IDs of the `PMM_PurchaseCandidate` which you skip here
  * use Loggables.withLogger(logger, Level.DEBUG) for that, to make sure it's in the async-workpackage-log as well as in the actually logback-log

File: de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java [hunk 3]
```diff
@@ -66,7 +64,20 @@ public Result processWorkPackage(final I_C_Queue_WorkPackage workPackage, final
 	private List<I_PMM_PurchaseCandidate> retrieveItems()
 	{
 		final I_C_Queue_WorkPackage workpackage = getC_Queue_WorkPackage();
-		return Services.get(IQueueDAO.class).retrieveItemsSkipMissing(workpackage, I_PMM_PurchaseCandidate.class, ITrx.TRXNAME_ThreadInherited);
+		return Services.get(IQueueDAO.class).retrieveAllItems(workpackage, I_PMM_PurchaseCandidate.class);
```
Comment (metas-ts) [inline de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java]: Already when enqeuting this, make sure to only add `PMM_PurchaseCandidate`s that have a quantity.
it's still possible theat they are zero-ed menwhile, but for that, there will be logging, see below

File: de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java [hunk 4]
```diff
@@ -68,19 +68,6 @@ public Result processWorkPackage(final I_C_Queue_WorkPackage workPackage, final
 		return Services.get(IQueueDAO.class).retrieveItemsSkipMissing(workpackage, I_PMM_PurchaseCandidate.class, ITrx.TRXNAME_ThreadInherited);
 	}
 
-	private List<I_PMM_PurchaseCandidate> getItemsWithQty(List<I_PMM_PurchaseCandidate> candidates)
```
Comment (metas-ts) [inline de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java]: alright, good you removed it again; `OrderLineAggregation` already does that job (i saw it only now 😊 ).

File: de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java [hunk 5]
```diff
@@ -1,16 +1,17 @@
...
-import org.adempiere.ad.trx.api.ITrx;
-import org.adempiere.exceptions.AdempiereException;
-
 import de.metas.async.api.IQueueDAO;
 import de.metas.async.model.I_C_Queue_WorkPackage;
 import de.metas.async.spi.WorkpackageProcessorAdapter;
 import de.metas.procurement.base.model.I_PMM_PurchaseCandidate;
 import de.metas.procurement.base.order.impl.OrdersGenerator;
 import de.metas.util.Services;
+import org.adempiere.ad.trx.api.ITrx;
```
Comment (metas-ts) [inline de.metas.procurement.base/src/main/java/de/metas/procurement/base/order/async/PMM_GenerateOrders.java:9]: please quickly clean up the imports; i think at least BigDecimal is not needed anymore

**Comment by metas-mk** [comment]
> one tiny request regarding imports..
> 
> PS: one thing i had forgotten: i think we need to put special care to the process name: almost every process besides this one works only on the selected items.
> So maybe we need to name it something like "Gefilterte zu Bestellungen" / "Filtered to pruchase orders"
> Wdyt?

Does this process only work on filtered records with order qty > 0 or for all records with order qty > 0 (no matter what is filtered).

I would prefer the process name "Bestellungen generieren für Mengen > 0" or "Create Purchase for Quantities > 0"

**Comment by dragospodariu96** [comment]
@metas-mk works just like in the swing client, taking all records currentyl in the view that have qtytoorder and qtyordered > 0. It takes into consideration all filters (like date, bpartner, organisation, etc) but regardless of how many records are manually highlighted, it filters all that are in the view at the moment the action is pressed.

**Pull request merged**